internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf(
+				"Expected *ProviderSourceData, got: %T. Please report this issue to the provider developers.",
+				req.ProviderData,
+			),
+		)
+
+		return
+	}
+
+	g.Data = data
+}
+
+func (g *GSI) ValidateConfig(
+	ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse,
+) {
+	var config providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if config.OrganizationId.IsUnknown() {
+		return
+	}
+
+	if !config.BuildIndexes.IsNull() {
+		if !config.IsPrimary.IsNull() ||
+			!config.IndexName.IsNull() ||
+			!config.IndexKeys.IsNull() ||
+			!config.Where.IsNull() ||
+			!config.PartitionBy.IsNull() {
+
+			resp.Diagnostics.AddAttributeError(
+				path.Root("build_indexes"),
+				"Invalid Attribute Configuration",
+				"build_indexes is set so other optional attributes must be null",
+			)
+			return
+		}
+	}
+
+	if !config.IsPrimary.ValueBool() {
+		if config.IndexName.ValueString() == "" {
+			resp.Diagnostics.AddAttributeError(
+				path.Root("index_name"),
+				"Missing Attribute Configuration",
+				"Expected index_name to be configured but is null",
+			)
+			return
+		}
+
+		if config.IndexKeys.IsNull() {
+			resp.Diagnostics.AddAttributeError(
+				path.Root("index_keys"),
+				"Missing Attribute Configuration",
+				"Expected index_keys to be configured but is null",
+			)
+			return
+		}
+
+	}
+
+	if config.IsPrimary.ValueBool() {
+		if !config.IndexKeys.IsNull() || !config.Where.IsNull() || !config.PartitionBy.IsNull() {
+			resp.Diagnostics.AddAttributeError(
+				path.Root("is_primary"),
+				"Invalid Attribute Configuration",
+				"A primary index cannot have index keys, where clause or partition by clause",
+			)
+			return
+		}
+	}
+}
+
+func (g *GSI) executeGsiDdl(ctx context.Context, plan *providerschema.GsiDefinition, ddl string) error {
+	url := fmt.Sprintf(
+		"%s/v4/organizations/%s/projects/%s/clusters/%s/queryService/indexes",
+		g.HostURL,
+		plan.OrganizationId.ValueString(),
+		plan.ProjectId.ValueString(),
+		plan.ClusterId.ValueString(),
+	)
+
+	cfg := api.EndpointCfg{Url: url, Method: http.MethodPost, SuccessStatus: http.StatusOK}
+	ddlRequest := api.IndexDDLRequest{Definition: ddl}
+	response, err := g.Client.ExecuteWithRetry(
+		ctx,
+		cfg,
+		ddlRequest,
+		g.Token,
+		nil,
+	)
+	if err != nil {
+		// Indexer returns an error if there is a build already in progress.
+		// It's not really an error because indexer will retry the build.
+		// This is problematic because if indexes are built in bulk the user
+		// will be spammed with errors.
+		//
+		// In this case we can ignore the error.
+		if apiError, ok := err.(*api.Error); ok {
+			if apiError.HttpStatusCode == http.StatusInternalServerError &&
+				strings.Contains(
+					strings.ToLower(apiError.Message), "build already in progress",
+				) {
+				return nil
+			}
+		}
```

Comment:
Is it possible for this error to occur when other indexes are being built and we are trying to perform another operation? Will it return the same error? If we return nil in this case, does it not give the user the perception that their operation was successful?

Reply:
My original intention was to poll the index but this is not a good approach.

Now provider will return an error.